contract_comm/currency/currency.go

```diff
@@ -94,97 +94,127 @@ const (
 )
 
 var (
-	cgExchangeRateNum = big.NewInt(1)
-	cgExchangeRateDen = big.NewInt(1)
-
 	medianRateFuncABI, _   = abi.JSON(strings.NewReader(medianRateABI))
 	balanceOfFuncABI, _    = abi.JSON(strings.NewReader(balanceOfABI))
 	getWhitelistFuncABI, _ = abi.JSON(strings.NewReader(getWhitelistABI))
 )
 
-type exchangeRate struct {
-	Numerator   *big.Int
-	Denominator *big.Int
+// ExchangeRate represent the exchangeRate for the pair (base, token)
+// Follows the equation: 1 base * ExchangeRate = X token
+type ExchangeRate struct {
+	numerator   *big.Int
+	denominator *big.Int
 }
 
-func ConvertToGold(val *big.Int, currencyFrom *common.Address) (*big.Int, error) {
-	if currencyFrom == nil {
-		return val, nil
+// NoopExchangeRate returns the noop rate
+// which results in a 1:1 conversion
+func NoopExchangeRate() *ExchangeRate { return nil }
+
+// MustNewExchangeRate creates an exchange rate, panic on error
+func MustNewExchangeRate(numerator *big.Int, denominator *big.Int) *ExchangeRate {
+	rate, err := NewExchangeRate(numerator, denominator)
+	if err != nil {
+		panic(err)
 	}
+	return rate
+}
 
-	celoGoldAddress, err := contract_comm.GetRegisteredAddress(params.GoldTokenRegistryId, nil, nil)
-	if err == errors.ErrSmartContractNotDeployed || err == errors.ErrRegistryContractNotDeployed {
-		log.Warn("Registry address lookup failed", "err", err)
-		return val, err
+// NewExchangeRate creates an exchange rate.
+// Requires numerator >=0 && denominator >= 0
+func NewExchangeRate(numerator *big.Int, denominator *big.Int) (*ExchangeRate, error) {
+	if numerator == nil || common.Big0.Cmp(numerator) >= 0 {
+		return nil, errors.ErrExchangeRateZero
 	}
+	if denominator == nil || common.Big0.Cmp(denominator) >= 0 {
+		return nil, errors.ErrExchangeRateZero
+	}
+	return &ExchangeRate{numerator, denominator}, nil
+}
 
-	if *currencyFrom == celoGoldAddress {
-		// This function shouldn't really be called with the token's address, but if it is the value
-		// is correct, so return nil as the error
-		return val, nil
-	} else if err != nil {
-		log.Error(err.Error())
-		return val, err
+// ToBase converts from token to base
+func (er *ExchangeRate) ToBase(tokenAmount *big.Int) *big.Int {
+	// check noop rate (cGLD rate)
+	if er == nil {
+		return new(big.Int).Set(tokenAmount)
 	}
-	return Convert(val, currencyFrom, nil)
+
+	return new(big.Int).Div(new(big.Int).Mul(tokenAmount, er.denominator), er.numerator)
```

Comment:
Note that changing the conversion logic in a way that would effect the way rounding errors are handled can be a hard fork if used consensus code. The two places I can think of that would be a risk are the cUSD validator rewards, which are converted to CELO and minted, and gas price minimum adherence where the gas value is converted to CELO before being compared to the GPM. Luckily in both cases, a non-CELO token is being converted to/from CELO, which makes this logic equivalent to the previous implementation. I'm not 100% certain there is not a place where this may cause issue. https://github.com/celo-org/celo-blockchain/blob/2dede5cce581dfe378f85c4c9d774ccd3ef99ac9/contract_comm/currency/currency.go#L148-L149

Reply:
My intention was to leave the exact same logic here. Minus the case of one of the rate being the 1/1 on which case i elminated the `*1` & `/1` steps. Which i though (maybe wrong) would just save a few cpu cycles, but be idential in behaviour.

Where do you see i've chaning the logic?

Reply:
What is changed is that when converting from one arbitrary currency to another (e.g. cUSD to cEUR), we would need to convert from source to the base (CELO) and then to the target currency, instead of converting from source to target in one operation. These are not guaranteed to get the same result due to a potential loss of precision in the conversion from source to base.

After having reviewed the whole PR, I realize there are indeed no places where we make a conversion from one non-CELO currency to another, so there are no places where the operation described above would need to be done. As a result, this change is fine.